docs/build/build-with-ai/build-with-goat/token-swap-agent.md

```diff
@@ -0,0 +1,233 @@
+      model: openai("gpt-4o-mini"),
+      tools: tools,
+      maxSteps: 10, // Maximum number of tool invocations per request
+      prompt: prompt,
+      onStepFinish: (event) => {
+        console.log(event.toolResults);
+      },
+    });
+
+    console.log("\n-------------------\n");
+    console.log("RESPONSE");
+    console.log("\n-------------------\n");
+    console.log(result.text);
+  } catch (error) {
+    console.error(error);
+  }
+  console.log("\n-------------------\n");
+}
+```
+
+This sets up the interactive loop where users can input natural language prompts to swap tokens:
+
+- The agent accepts commands like "Swap cUSD for CELO" or "Check my token balance"
+- `generateText()` processes these commands through the AI model (gpt-4o-mini)
+- The AI determines which tools to call to fulfill the request
+- Results are returned in natural language
+
+## Setting Up Your Celo Token Swap Agent
+
+Follow these steps to set up your own token swap agent on Celo:
+
+### 1. Clone the Repository
+
+```bash
+git clone https://github.com/goat-sdk/goat.git && cd goat
+```
+
+### 2. Install and Build Dependencies
+
+```bash
+cd typescript
+pnpm install
+pnpm build
+```
+
+### 3. Navigate to the Example Directory
+
+```bash
+cd examples/by-use-case/evm-swap-tokens
+```
+
+### 4. Configure Environment Variables
+
+Create a `.env` file with your credentials:
+
+```bash
+cp .env.template .env
+```
+
+Edit the `.env` file to include:
+
+```plaintext
+OPENAI_API_KEY=your_openai_api_key
+WALLET_PRIVATE_KEY=your_wallet_private_key
+RPC_PROVIDER_URL=https://forno.celo.org  # Celo Mainnet
+# or use https://alfajores-forno.celo-testnet.org for Testnet
+UNISWAP_BASE_URL=provided_value  # Will be populated from template
+UNISWAP_API_KEY=provided_value   # Will be populated from template
+```
+
+For production use, you can get your own Uniswap API key from [Uniswap Hub](https://www.uniswap.org/developers).
+
+### 5. Run the Interactive CLI
+
+```bash
+pnpm ts-node index.ts
+```
+
+## Using the Token Swap Agent on Celo
+
+Once your agent is running, you can interact with it using natural language. Here are some examples specifically for Celo tokens:
+
+```
+Swap cUSD for CELO
+Swap CELO for cUSD
+Check my token balance
+What's the current exchange rate between cUSD and CELO?
+```
+
+The agent will interpret your commands, call the appropriate tools, and execute the token swaps on your behalf.
+
+## Production Considerations for Token Swap Agents
+
+When deploying token swap agents in production environments, consider using smart wallets to:
+
+### Enhance Security with Programmable Permissions
+
+- Limit fund amounts for transactions
+- Restrict contract interactions to trusted protocols
+- Define required signatures for high-value transactions
+
+### Maintain Regulatory Compliance with Non-Custodial Wallets
+
+- Ensure agent platforms never have access to users' wallets
+- Avoid money transmitter license requirements
+- Improve user trust and control
+
```

Comment:
do you have any resources where people can read up on this? I would not know how to do this tbh